### ISSUE TITLE
Terraform Test: helper module data sources returns error while tearing down

### ISSUE THREAD

**Issue by maciejr2** [hashicorp/terraform#34280]
### Terraform Version

```shell
Terraform v1.6.4
on windows_amd64
+ provider registry.terraform.io/hashicorp/local v2.4.0
```

### Terraform Configuration Files

```terraform
# main.tf
terraform {
  required_version = ">= 0.12.26"
}

resource "local_file" "example" {
  content  = var.file_content
  filename = var.file_name
}

variable "file_name" {
  type = string
}
variable "file_content" {
  type = string
}
```

```terraform
# testing/verify/main.tf
data "local_file" "foo" {
  filename = var.file_name
}
variable "file_name" {
  type = string
}
```

```terraform
# tests/file_test.tftest.hcl
variables {
    file_name = "example.txt"
    file_content = "test"
}

run "test_example" {
}

run "verify" {
    module {
        source = "./testing/verify"
    }

    assert {
        condition = data.local_file.foo.content == "test"
        error_message = "wrong file content"
    }
}
```

### Debug Output

FULL LOG https://gist.github.com/maciejr2/b0126d6d5168e32f2d7db2893ffb5e4e

```
Terraform encountered an error destroying resources created while executing
tests\file_test.tftest.hcl/verify.

Error: Read local file data source error

  with data.local_file.foo,
  on testing\verify\main.tf line 1, in data "local_file" "foo":
   1: data "local_file" "foo" {

The file at given path cannot be read.

+Original Error: open example.txt: The system cannot find the file specified.
tests\file_test.tftest.hcl... fail

Failure! 2 passed, 0 failed.
```
### Expected Behavior

Data sources from helper modules should be destroyed before main state configuration to not throw an error.

### Actual Behavior

It looks like data resources are not destroyed before main resources are removed which leads to errors. 
It tries to destroy data source when file resource does not exist anymore.

### Steps to Reproduce

1. `terraform init`
2. `terraform test`

### Additional Context

This issue occurred while tearing down testing helper module with  `data "aws_s3_objects" "this"`. 
Tests are passing, but tearing down produces errors.
I reproduced it using `local` provider.

**Comment by liamcervante** [comment]
Hi @maciejr2, thanks for filing this. I can replicate it so I've marked it as confirmed.

I'll take a look and see what's happening here. I'd have thought trying to destroy a data source that doesn't exist should work, but it looks like it's attempting to read it before destroying it which is producing the error. Maybe we can tell the destroy operations in tests to skip refreshes, or we might have to rearrange the order of destroys as you've suggested.

**Comment by liamcervante** [comment]
This will be fixed in the upcoming v1.7 release. The testing framework will destroy the run blocks in simple reverse block order instead of trying to overcomplicate things. I think this will address the specific use case here, as the data sources will no longer try to load after the main state has been destroyed.

I'll update the documentation for v1.6 separately to clarify the behaviour and pitfalls here.

Thanks!
